test-tool/test_async_write.c

```diff
@@ -97,12 +106,91 @@ test_async_write(void)
 					       &state);
 		CU_ASSERT_EQUAL(ret, 0);
 
-		state.dispatched++;
+		state.io_dispatched++;
+		logging(LOG_VERBOSE, "WRITE10 dispatched: %d of %d (cmdsn=%d)",
+			state.io_dispatched, num_ios, atask->cmdsn);
+	}
+
+	while (state.io_completed < state.io_dispatched) {
+		struct pollfd pfd;
+
+		pfd.fd = iscsi_get_fd(sd->iscsi_ctx);
+		pfd.events = iscsi_which_events(sd->iscsi_ctx);
+
+		ret = poll(&pfd, 1, -1);
+		CU_ASSERT_NOT_EQUAL(ret, -1);
+
+		ret = iscsi_service(sd->iscsi_ctx, pfd.revents);
+		CU_ASSERT_EQUAL(ret, 0);
+	}
+
+	free(buf);
+}
+
+static void
+test_async_io_logout_cb(struct iscsi_context *iscsi __attribute__((unused)),
+			int status, void *command_data __attribute__((unused)),
+			void *private_data)
+{
+	struct tests_async_write_state *state = private_data;
+
+	state->logout_cmpl++;
+	logging(LOG_VERBOSE, "Logout completed with %d IOs outstanding",
+		state->io_dispatched - state->io_completed);
+	CU_ASSERT_EQUAL(status, SCSI_STATUS_GOOD);
+}
+
+void
+test_async_io_logout(void)
+{
+	int i, ret;
+	struct tests_async_write_state state = { 0, 0, 0, 0, 0 };
+	int blocks_per_io = 8;
+	int num_ios = 10;
+	/* IOs in flight concurrently, but all using the same src buffer */
+	unsigned char *buf;
+
+	CHECK_FOR_DATALOSS;
+	CHECK_FOR_SBC;
+	CHECK_FOR_ISCSI(sd);
+
+	if (maximum_transfer_length
+	 && (maximum_transfer_length < (blocks_per_io * num_ios))) {
```

Comment:
Do you agree that the parentheses can be left out from this expression without changing the meaning of this expression and without reducing readability of this code?

Reply:
I'd prefer to keep the brackets around the multiplication, if that's okay with you. I'll drop the others.